ENH: ``meson`` backend for ``f2py``

HaoZeke opened the pull request:
This distills and extends #22225 where @NamamiShanker completed the design and implementation of the CLI refactor and backend work as his GSoC project last year.

This uses the `f2py2e` CLI instead of the `f2pyarg` CLI.

Closes #24444. Closes #22910. Closes #21484. Closes #20240.
Actually pretty much [closes out every](https://github.com/numpy/numpy/labels/component%3A%20numpy.distutils) `np.distutils` issue.
Part of #23981. Should be good to go for 1.26.

It is **very** minimal, and warnings have been added. That said this:

- Adds an abstract backend
- Shoehorns Distutils into said backend
- Adds a meson builder (optional) backend

Usage, for the standard documentation example of:

```fortran
C FILE: FIB1.F
      SUBROUTINE FIB(A,N)
C
C     CALCULATE FIRST N FIBONACCI NUMBERS
C
      INTEGER N
      REAL*8 A(N)
      DO I=1,N
         IF (I.EQ.1) THEN
            A(I) = 0.0D0
         ELSEIF (I.EQ.2) THEN
            A(I) = 1.0D0
         ELSE
            A(I) = A(I-1) + A(I-2)
         ENDIF
      ENDDO
      END
C END FILE FIB1.F
```

Which has been tested with:

```python
# Meson, will only generate the .so file
python -m numpy.f2py -c --f90flags='-O3' -m fib fib.f --backend meson
# Meson, will generate the .so, leave a build directory
python -m numpy.f2py -c --f90flags='-O3' -m fib fib.f --backend meson --build-dir blah
# Test it
python -c "import numpy as np; import fib as fibby; a = np.zeros(9); fibby.fib(a); print (a)"
```

Also tested with a corresponding `.pyf`:
```fortran
python module fib2 ! in 
    interface  ! in :fib2
        subroutine fib(a,n) ! in :fib2:fib.f
            real*8 dimension(n) :: a
            integer, optional,check(shape(a, 0) == n),depend(a) :: n=shape(a, 0)
        end subroutine fib
    end interface 
end python module fib2
```
Which works as expected:
```bash
python -m numpy.f2py -c --f90flags='-O3' fib1.pyf fib.f --debug --backend meson
python -c "import numpy as np; import fib2 as fibby; a = np.zeros(9); fibby.fib(a); print (a)"
[ 0.  1.  1.  2.  3.  5.  8. 13. 21.]
```

There's also a new `--dep` flag which can be passed (multiple times if necessary) to use `f2py` for more than trivial use cases:

```bash
f2py -c --backend meson GaussJacobiQuadCCPy.pyf ../src/gjp_lapack.f90 - -dep lapack 
```

Which makes it IMO much easier to use than the older backend.

The default backend is `distutils` on Python versions '<3.12.0' (though it will emit a warning).

## Required
Do not merge until these are finished!!!
- [X] Works with `.f` and `.f90`
- [X] Works with `.pyf`
- [X] Testsuite passes (see next section)

## `python-3.12` Support

Personally I use `pyenv`:
```bash
pyenv install 3.12.0rc1
eval "$(pyenv init -)"
pyenv global 3.12.0rc1
pip install -r build_requirements.txt
pip install -r test_requirements.txt
spin run $SHELL # manual tests
spin test -t numpy.f2py -m full
```

The tests pass manually (see exceptions below), some changes are needed there to be more flexible. For now, this patch can be used:
```diff
diff --git i/numpy/f2py/f2py2e.py w/numpy/f2py/f2py2e.py
index 6b63a6c6b..fad5db12f 100755
--- i/numpy/f2py/f2py2e.py
+++ w/numpy/f2py/f2py2e.py
@@ -597,7 +597,7 @@ def run_compile():
         backend_key = sys.argv.pop(backend_index + 1)
         sys.argv.pop(backend_index)
     else:
-        backend_key = 'distutils'
+        backend_key = 'meson'
     build_backend = f2py_build_generator(backend_key)
 
     modulename = 'untitled'
diff --git i/numpy/f2py/tests/util.py w/numpy/f2py/tests/util.py
index 75b257cdb..abb281d7b 100644
--- i/numpy/f2py/tests/util.py
+++ w/numpy/f2py/tests/util.py
@@ -370,10 +370,6 @@ def setup_method(self):
         if self.module is not None:
             return
 
-        # Check compiler availability first
-        if not has_c_compiler():
-            pytest.skip("No C compiler available")
-
         codes = []
         if self.sources:
             codes.extend(self.sources)
@@ -390,12 +386,6 @@ def setup_method(self):
                 needs_f90 = True
             elif str(fn).endswith(".pyf"):
                 needs_pyf = True
-        if needs_f77 and not has_f77_compiler():
-            pytest.skip("No Fortran 77 compiler available")
-        if needs_f90 and not has_f90_compiler():
-            pytest.skip("No Fortran 90 compiler available")
-        if needs_pyf and not (has_f90_compiler() or has_f77_compiler()):
-            pytest.skip("No Fortran compiler available")
 
         # Build the module
         if self.code is not None:
```

### Exceptions
- `test_array_from_pyobj` will be skipped for now since it uses a bunch of `distutils` helper functions
- `TestCReturnReal` fails for now, though manual testing of the code works, has to do with the way modules are built for testing

#### Manual `CReturnReal` setup

Consider:

```fortran
python module c_ext_return_real
usercode '''
float t4(float value) { return value; }
void s4(float *t4, float value) { *t4 = value; }
double t8(double value) { return value; }
void s8(double *t8, double value) { *t8 = value; }
'''
interface
  function t4(value)
    real*4 intent(c) :: t4,value
  end
  function t8(value)
    real*8 intent(c) :: t8,value
  end
  subroutine s4(t4,value)
    intent(c) s4
    real*4 intent(out) :: t4
    real*4 intent(c) :: value
  end
  subroutine s8(t8,value)
    intent(c) s8
    real*8 intent(out) :: t8
    real*8 intent(c) :: value
  end
end interface
end python module c_ext_return_real
```

Which is just the test-case extracted into a file. This does have the attributes tested:

```bash
python -m numpy.f2py -c c_ext_return_real.pyf --backend meson
python -c "import c_ext_return_real as cextrr; print(dir(cextrr)[9:])"
```

It should run in the test suite once the utilities there have been reworked to use the abstract `Backend` class instead.

## Post PR
### High Priority
- The helpers for the `f2py` test suite are still very `distutils` specific, and need to be updated to use the `Backend` class.
### Medium Priority
- Update the User Guide with examples, pedagogical explanations, pithy comments, etc.
### Lower Priority
This ignores every `-c` flag except `build-dir` and warns the user to use the skeleton `meson.build` as a baseline to which should be edited. In any case though the `-c` flags need to be rethought for `meson` and there's already been excellent work in that direction last year (#22225). **Later**, when the `f2pyarg` work lands, or otherwise, this can be extended.

NamamiShanker committed:
FIX: Import f2py2e rather than f2py for run_main

NamamiShanker committed:
FIX: Import f2py2e instead of f2py

HaoZeke committed:
ENH: Add F2PY back-end work from gh-22225

Co-authored-by: NamamiShanker <NamamiShanker@users.noreply.github.com>

HaoZeke committed:
ENH: Add meson skeleton from gh-2225

Co-authored-by: NamamiShanker <NamamiShanker@users.noreply.github.com>

HaoZeke committed:
MAINT: Trim backend.py down to f2py2e flags

HaoZeke committed:
ENH: Add a factory function for backends

HaoZeke committed:
ENH: Add a distutils backend

HaoZeke committed:
ENH: Handle --backends in f2py

Defaults to distutils for now

HaoZeke committed:
DOC: Add some minor comments in f2py2e

HaoZeke committed:
MAINT: Refactor and rework meson.build.src

HaoZeke committed:
MAINT: Add objects

HaoZeke committed:
MAINT: Cleanup distutils backend

HaoZeke committed:
MAINT: Refactor to add everything back to backend

Necessary for the meson.build for now. Refactors / cleanup needs better
argument handling in f2py2e

HaoZeke committed:
MAINT: Fix overly long line

HaoZeke committed:
BUG: Construct wrappers for meson backend

HaoZeke committed:
MAINT: Rework, simplify template massively

HaoZeke committed:
ENH: Truncate meson.build to skeleton only

HaoZeke committed:
MAINT: Minor backend housekeeping, name changes

HaoZeke committed:
MAINT: Less absolute paths, update setup.py [f2py]

HaoZeke committed:
MAINT: Move f2py module name functionality

Previously part of np.distutils

HaoZeke committed:
ENH: Handle .pyf files

HaoZeke committed:
TST: Fix typo in isoFortranEnvMap.f90

HaoZeke committed:
MAINT: Typo in f2py2e support for pyf files

HaoZeke committed:
DOC: Add release note for --backend

HaoZeke committed:
MAINT: Conditional switch for Python 3.12 [f2py]

HaoZeke commented:
@rgommers, @mattip, @charris (everyone else), I think this should be good to go in as is, it encapsulates the main user-facing features and works on `3.12`. The changes to the testsuite (in order to use `Backend`) should be in a followup, to prevent scope creep and make things easier to review. If there's a very strong reason to have the testsuite changes in this PR I could add them, otherwise I'd prefer to do it in a follow up later. Similar considerations apply to changing the user documentation.

rgommers commented:
Great to hear that! I'll try to review within the next couple of days.

HaoZeke committed:
MAINT: No absolute paths in backend [f2py-meson]

The files are copied over anyway, this makes it easier to extend the
generated skeleton

HaoZeke committed:
MAINT: Prettier generated meson.build files [f2py]

HaoZeke committed:
ENH: Add meson's dependency(blah) to f2py

HaoZeke committed:
DOC: Document the new flag

rgommers commented on numpy/f2py/_backends/_meson.py:
```diff
@@ -0,0 +1,156 @@
+    ):
+        self.modulename = modulename
+        self.build_template_path = (
+            Path(__file__).parent.absolute() / "src" / "meson.build.src"
```
Can you choose another file extension name? `.src` has a specific meaning in numpy (see all the `.c.src` files) and this has nothing to do with that. `.template` or some such thing will do here.

rgommers commented on numpy/f2py/_backends/_meson.py:
```diff
@@ -0,0 +1,156 @@
+class MesonBackend(Backend):
+    def __init__(self, *args, **kwargs):
+        warnings.warn(
+            "The MesonBackend is experimental and may change in the future."
```
`MesonBackend` -> `Meson backend`

rgommers replied:
More importantly, if this is the default for Python 3.12, it should not emit any warnings. You can just delete it. If there's a bug, it can be fixed normally - that should not affect the user's `f2py -c` invocation.

rgommers commented on numpy/f2py/_backends/src/meson.build.src:
```diff
@@ -0,0 +1,40 @@
+project('${modulename}',
```
Besides the file name, it'd be better not to have an extra subdir here with only a single file in it. You can move it up.

rgommers commented on numpy/f2py/_backends/src/meson.build.src:
```diff
@@ -0,0 +1,40 @@
+        ['c', 'fortran'],
+        version : '0.1',
+        meson_version: '>= 1.1.0',
+        default_options : ['warning_level=2'])
```
Is this necessary? A higher warning level is typically not desired I think.

rgommers replied:
Also, did you think about what the default build type should be? Perhaps it should be `release`?

HaoZeke replied:
> Is this necessary? A higher warning level is typically not desired I think.

Ah good point, I see the [default is](https://mesonbuild.com/Builtin-options.html#core-options) now $1$, when I wrote the `f2py` with `meson` [documentation](https://numpy.org/doc/stable/f2py/buildtools/meson.html) it defaulted to $3$ so this was a lowering done to stop passing `-fimplicit-none`. I'll explicitly set it to $1$ then, just in case the defaults change later.

HaoZeke replied:
> Also, did you think about what the default build type should be? Perhaps it should be `release`?

It should respect `f2py`'s `--debug` flag, updating.

rgommers commented on numpy/f2py/_backends/src/meson.build.src:
```diff
@@ -0,0 +1,40 @@
+        default_options : ['warning_level=2'])
+
+py_mod = import('python')
+py = py_mod.find_installation(pure: false)
```
Minor: newer practice is to put this on a single line because `py_mod` isn't needed for anything:
```meson
py = import('python').find_installation(pure: false)
```

rgommers commented on numpy/f2py/f2py2e.py:
```diff
@@ -141,6 +143,19 @@
+                       containing ["lapack", "scalapack"].
+
+                       Note: The trailing `:' is required to signify the end of
+                       the dependency list. This is only relevant for the meson backend.
```
Can you explain why this is needed only for Meson?

rgommers replied:
Oh wait, it is mentioned in the PR description. It's a new feature, not a necessary change. It seems okay to me to add, although at some point using `f2py -c` is a bit questionable for non-trivial things of course. The syntax seems unusual though. Wouldn't it be better to have a regular `--dep` that takes a regular string name? And then if you have multiple, just repeat `--dep` for your next dependencies? That would avoid the odd trailing colon, and this is similar to `--config-settings` for `pip` for example.

HaoZeke replied:
This was implemented this way in keeping with the existing `skip:` and `only:` commands, and because it seemed easier when directly parsing `sys.argv`. Though its much better to use more modern styles, I'll add an `argparse` pre-parse for `sys.argv` to handle `--dep` :)

rgommers commented on numpy/f2py/f2py2e.py:
```diff
@@ -699,6 +724,12 @@ def main():
+                      'first was largely handled by distutils which has been removed in Python 3.12.'
+                      'The meson replacement is a work in progress, consider running without -c.'
+                      'Recall that the build can be customized with --build-dir',
+                      stacklevel=2)
```
Same comment as before: the default behavior of `f2py -c` should not emit warnings. Is this needed?

rgommers reviewed (commented):
Thanks @HaoZeke! Looks good overall. I did a first review and noted things that stood out to me. I haven't done actual testing yet.

HaoZeke committed:
MAINT: Simplify and rename backend template [f2py]

Co-authored-by: rgommers <rgommers@users.noreply.github.com>

HaoZeke committed:
ENH: Support build_type via --debug [f2py-meson]

HaoZeke committed:
MAINT,DOC: Reduce warn,rework doc [f2py-meson]

Co-authored-by: rgommers <rgommers@users.noreply.github.com>

HaoZeke committed:
ENH: Rework deps: to --dep calls [f2py-meson]

Also shows how incremental updates to the parser can be done.

HaoZeke committed:
MAINT,DOC: Add --backend to argparse, add docs

HaoZeke committed:
MAINT: Rename meson template [f2py-meson]

HaoZeke committed:
MAINT: Add meson.build for f2py

Should address https://github.com/numpy/numpy/pull/22225#issuecomment-1697208937

charris commented:
Needs rebase.

HaoZeke commented:
> Needs rebase.

yup, done :)

rgommers commented on numpy/meson.build:
```diff
@@ -413,3 +413,4 @@ subdir('core')
 subdir('fft')
 subdir('linalg')
 subdir('random')
+subdir('f2py')
```
This is duplicate, `f2py` is already listed as a pure Python module higher up.

rgommers commented on numpy/f2py/meson.build:
```diff
@@ -0,0 +1,14 @@
+    '_backends/meson.build.template',
+  ],
+  subdir: 'numpy/f2py/_backends'
+)
```
This file isn't needed.

rgommers committed:
BLD: remove duplicate f2py handling in meson.build files

rgommers commented on numpy/f2py/_backends/meson.build.template:
```diff
@@ -0,0 +1,42 @@
+project('${modulename}',
```
The formatting in this file is a little off - may be worth fixing in a next PR. Normal indentation is 2 or 4 spaces for function calls like `project(...)` or `py.extension_module(...)`.

rgommers reviewed (approved):
This looks good as far as I can tell from a line-by-line review and some initial testing with both small Fortran files and with SciPy (for regression purposes, it does not use `f2py -c`).

There is one issue in the `meson.build` config that I commented on, will push a fix for that. Then, once CI comes back green, I think this can be squash-merged and backported to `1.26.x`.

charris commented:
Thanks Rohit and Ralf.

rgommers commented:
Great to get this in - nice work @HaoZeke and @NamamiShanker!
